solutionbox/structured_data/test_mltoolbox/test_package_functions.py

```diff
@@ -0,0 +1,139 @@
+            file_pattern=['file1.txt'],
+            schema='col1:STRING,col2:INTEGER,col3:FLOAT'),
+        project_id='project_id').wait()
+    self.assertIn('project_id only needed if cloud is True',
+                  job.fatal_error.message)
+
+  def test_cloud_with_local_output_folder(self):
+    job = core_sd.analyze_async(
+        'some_dir',
+        dlml.CsvDataSet(
+            file_pattern=['gs://file1.txt'],
+            schema='col1:STRING,col2:INTEGER,col3:FLOAT'),
+        project_id='project_id',
+        cloud=True).wait()
+    self.assertIn('File some_dir is not a gcs path', job.fatal_error.message)
+
+  def test_cloud_but_local_files(self):
+    job = core_sd.analyze_async(
+        'gs://some_dir',
+        dlml.CsvDataSet(
+            file_pattern=['file1.txt'],
+            schema='col1:STRING,col2:INTEGER,col3:FLOAT'),
+        project_id='project_id',
+        cloud=True).wait()
+    self.assertIn('File file1.txt is not a gcs path', job.fatal_error.message)
+
+  def test_unsupported_schema(self):
+    """Test supported schema values.
+
+    Note that not all valid BQ schema values are valid/used in the structured
+    data package
+    """
+
+    unsupported_col_types = ['bytes', 'boolean', 'timestamp', 'date', 'time',
+                             'datetime', 'record']
+    for col_type in unsupported_col_types:
+      schema = 'col_name:%s' % col_type
+
+      job = core_sd.analyze_async(
+        'some_dir',
+        dlml.CsvDataSet(
+            file_pattern=['file1.txt'],
+            schema=schema),
+        cloud=False).wait()
+      self.assertIn('Schema contains an unsupported type %s.' % col_type,
+                    job.fatal_error.message)
+
+      job = core_sd.analyze_async(
+        'gs://some_dir',
+        dlml.CsvDataSet(
+            file_pattern=['gs://file1.txt'],
+            schema=schema),
+        cloud=True).wait()
+      self.assertIn('Schema contains an unsupported type %s.' % col_type,
+                    job.fatal_error.message)
+
+
+class TestFunctionSignature(unittest.TestCase):
+
+  def test_same_analysis(self):
+    """Test that there is only one analyze function"""
+    self.assertIs(core_sd.analyze, classlin.analyze)
+    self.assertIs(core_sd.analyze, classdnn.analyze)
+    self.assertIs(core_sd.analyze, reglin.analyze)
+    self.assertIs(core_sd.analyze, regdnn.analyze)
+
+  def test_same_analysis_async(self):
+    """Test that there is only one analyze_async function"""
+    self.assertIs(core_sd.analyze_async, classlin.analyze_async)
+    self.assertIs(core_sd.analyze_async, classdnn.analyze_async)
+    self.assertIs(core_sd.analyze_async, reglin.analyze_async)
+    self.assertIs(core_sd.analyze_async, regdnn.analyze_async)
+
+  def test_analysis_argspec(self):
+    """Test all analyze functions have the same parameters"""
+    if sys.version_info.major > 2:
+      spec_fn = inspect.getfullargspec  # for python 3 compatibility
+    else:
+      spec_fn = inspect.getargspec
```

Comment:
Do you think it could be helpful to have a compatibility module pyDatalab? The compatibility module could contain helper function for common python 2/3 compatibility issues. For example:

```
def compat_inspect_getargspec(obj):
	if sys.version_info.major > 2:
		return obj.getfullargspec  # for python 3 compatibility
	else:
		return obj.getargspec(**kw)
```

Then we can use the compatibility functions directly:
```
    self.assertEqual(spec_fn(core_sd.analyze), compat_inspect_getargspec(core_sd.analyze_async))
    self.assertEqual(spec_fn(core_sd.analyze), compat_inspect_getargspec(core_sd._analyze))
```